.ci/scripts/dswx_compare_opera_pge.py

```diff
@@ -0,0 +1,221 @@
+from osgeo import gdal
+import numpy as np
+import os
+import sys
+
+COMPARE_DSWX_HLS_PRODUCTS_ERROR_TOLERANCE_ATOL = 1e-6
+COMPARE_DSWX_HLS_PRODUCTS_ERROR_TOLERANCE_RTOL = 1e-5
+
+def _get_prefix_str(flag_same, flag_all_ok):
+    flag_all_ok[0] = flag_all_ok[0] and flag_same
+    return '[OK]   ' if flag_same else '[FAIL] '
+
+
+def compare_dswx_hls_products(file_1, file_2):
+    if not os.path.isfile(file_1):
+        print(f'ERROR file not found: {file_1}')
+        return False
+
+    if not os.path.isfile(file_2):
+        print(f'ERROR file not found: {file_2}')
+        return False
+
+    flag_all_ok = [True]
+
+    # TODO: compare projections ds.GetProjection()
+    layer_gdal_dataset_1 = gdal.Open(file_1, gdal.GA_ReadOnly)
+    geotransform_1 = layer_gdal_dataset_1.GetGeoTransform()
+    metadata_1 = layer_gdal_dataset_1.GetMetadata()
+    nbands_1 = layer_gdal_dataset_1.RasterCount
+
+    layer_gdal_dataset_2 = gdal.Open(file_2, gdal.GA_ReadOnly)
+    geotransform_2 = layer_gdal_dataset_2.GetGeoTransform()
+    metadata_2 = layer_gdal_dataset_2.GetMetadata()
+    nbands_2 = layer_gdal_dataset_2.RasterCount
+
+    # compare number of bands
+    flag_same_nbands =  nbands_1 == nbands_2
+    flag_same_nbands_str = _get_prefix_str(flag_same_nbands, flag_all_ok)
+    prefix = ' ' * 7
+    print(f'{flag_same_nbands_str}Comparing number of bands')
+    if not flag_same_nbands:
+        print(prefix + f'Input 1 has {nbands_1} bands and input 2'
+              f' has {nbands_2} bands')
+        return False
+
+    # compare array values
+    print('Comparing DSWx bands...')
+    total_image_differences = 0
+    min_difference = 0
+    max_difference = 0
+    for b in range(1, nbands_1 + 1):
+        gdal_band_1 = layer_gdal_dataset_1.GetRasterBand(b)
+        gdal_band_2 = layer_gdal_dataset_2.GetRasterBand(b)
+        image_1 = gdal_band_1.ReadAsArray()
+        image_2 = gdal_band_2.ReadAsArray()
+        is_close_result = np.isclose(image_1, image_2,
+                                     atol=COMPARE_DSWX_HLS_PRODUCTS_ERROR_TOLERANCE_ATOL,
+                                     rtol=COMPARE_DSWX_HLS_PRODUCTS_ERROR_TOLERANCE_RTOL)
+        num_image_differences = np.sum(~is_close_result)
+        if num_image_differences > 0:
+            flag_bands_are_equal = False
+            # Find the min and max differences using the numpy.isclose function
+            # 0 <= (atol + rtol * absolute(b)) - absolute(a-b)
+            a1 = image_1[~is_close_result]
+            b2 = image_2[~is_close_result]
+            atol = COMPARE_DSWX_HLS_PRODUCTS_ERROR_TOLERANCE_ATOL
+            rtol = COMPARE_DSWX_HLS_PRODUCTS_ERROR_TOLERANCE_RTOL
+            is_close_differences = abs((atol + rtol * abs(b2)) - abs(a1 - b2))
```

Comment:
Is it possible that somebody would want to see the exact values that fell outside of the numpy.isclose() calculation in order to adjust the rtol and atol values and adjust the pass/fail tolerance of the comparison?

Reply:
Not just possible, but highly likely that someone would do such an analysis.  Though the purpose should be on fixing the algorithm, rather than adjusting our criteria to get a passing test.  And that's easy enough to do offline, outside this script.

The main thing I was pointing out here is that, as written, `is_close_differences` aren't the actual differences, since the first term on the RHS adds a small value to the absolute differences.  You already know where the arrays are different, so just use that, rather than redoing the calculation behind `np.isclose()`.

Looking back at the code, and my suggestion up above, I might have gotten the meaning of `is_close_result` inverted.  Use the one that picks out where they're different (either `is_close_result` or `~is_close_result`).